### PR TITLE
Delete / Publish / Unpublish actions on instructorHome redirects to Feedbacks page instead of Home. #3649

### LINKED ISSUE #3649: Delete / Publish / Unpublish actions on instructorHome redirects to Feedbacks page instead of Home.
As mentioned in the title.
Caused by the line below found in instructorHome.jsp.
`<%=data.getInstructorFeedbackSessionActions(fdb, false, instructor, courseIdSectionNamesMap.get(fdb.courseId))%>`

The comment and signature for the method is below.
````
* @param isHome
*        Flag whether the link is to be put at homepage (to determine the redirect link in delete / publish)
getInstructorFeedbackSessionActions(FeedbackSessionAttributes session, boolean isHome, InstructorAttributes instructor, List<String> sectionsInCourse)
````
Is this expected behaviour?

### PR DESCRIPTION
**Pull request by kennho**
Fixes #3649

### CODE REVIEW

**Comment by kennho** [comment]
I was scrolling through the issues and I found #3360 which looks like a duplicate / subset of this? I also saw #3400 which is similar to this.

Actually, I have changed the "copy" and the "remind" (other than "publish / unpublish" and "delete") button on instructorHome to redirect back to home too. Should I remove the change and submit a pull request under #3400 instead?

File: src/main/java/teammates/ui/controller/InstructorFeedbackDeleteAction.java [hunk 1]
```diff
@@ -29,7 +32,9 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         statusToAdmin = "Feedback Session <span class=\"bold\">[" + feedbackSessionName + "]</span> "
                         + "from Course: <span class=\"bold\">[" + courseId + " deleted.";
         
-        return createRedirectResult(Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE);
+        nextUrl = nextUrl == null ? Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE : nextUrl;
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackDeleteAction.java]: This seems to be repeated above?

File: src/main/java/teammates/ui/controller/PageData.java [hunk 2]
```diff
@@ -525,11 +539,20 @@ public String getInstructorStudentCommentClearPendingLink(String courseId) {
...
     
-    public String getInstructorFeedbackDeleteLink(String courseId, String feedbackSessionName, String nextURL) {
+    /**
+     * Retrieves the link to submit request to delete the session.
+     * @param courseId course ID
+     * @param feedbackSessionName the session name
+     * @param isHome true to indicate the request is from the home page
+     * @return the link to submit request to delete the session with return page link
+     */
+    public String getInstructorFeedbackDeleteLink(String courseId, String feedbackSessionName, boolean isHome) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/PageData.java]: Why change from using <code>nextUrl</code> to <code>isHome</code>?

**Comment by kanghj** [comment]
@kennho, I think it would be better if this was done under separate issues, since this pr is quite big. But I'll leave the choice to you if you want to split this up into different issues or continue with what you already have here.

File: src/main/java/teammates/ui/controller/PageData.java (hunk 2 above)
Comment (kennho) [inline src/main/java/teammates/ui/controller/PageData.java]: oh because previously the buildBasicFsForm was passing an empty string inside. This is causing an error after I updated InstructorFeedbackDeleteAction to use nextUrl as nextUrl will give me something like `*nextUrl=user=text@example.com` 
Should I change it to take in String then pass in the correct String when calling instead?

Comment (kanghj) [inline src/main/java/teammates/ui/controller/PageData.java]: Yes, I think it's better to pass in the correct String. We can have a default nextUrl if it's empty

Comment (kennho) [inline src/main/java/teammates/ui/controller/PageData.java]: Okay. Do I do this for getInstructorFeedbackDeleteLink only or for the rest of the similar methods also? @kanghj

Comment (kanghj) [inline src/main/java/teammates/ui/controller/PageData.java]: Let's do it for the similar methods that were modified in this pr.

**Comment by shreyamahapatra** [comment]
people..I have cloned the repository..followed all the step but could not access admin/adminHomePage.please help me out

**Comment by thyageshm** [comment]
@shreyamahapatra please create a new issue... I will be removing your comment here because it is not relevant to this thread.

**Comment by kennho** [comment]
Updated with suggested changes for delete, publish and unpublish link. Removed the fix to the "copy" button and submitted another PR on issue #3400 for the "copy" button. Since this PR is to solve the delete, publish and unpublish button, I am going to remove the fix to the "remind" button and create another issue on it. @kanghj

**Comment by kennho** [comment]
Removed the fix to the "remind" button. Created issue #5088 and submitted a PR. Also updated methods to take in return url instead of taking in isHome. @kanghj

File: src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java [hunk 3]
```diff
@@ -146,7 +146,7 @@ private void addAttributeIf(boolean shouldAdd, ElementTag button, String key, St
                     getInstructorFeedbackStatsLink(session.courseId, session.feedbackSessionName),
                     isRecent,
                     getInstructorFeedbackSessionActions(
-                            session, false, instructor));
+                            session, true, instructor));
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java]: Can you comment here why this change is made?

File: src/main/java/teammates/ui/controller/PageData.java [hunk 4]
```diff
@@ -525,12 +525,20 @@ public String getInstructorStudentCommentClearPendingLink(String courseId) {
         return link;
     }
     
-    public String getInstructorFeedbackDeleteLink(String courseId, String feedbackSessionName, String nextURL) {
+    /**
+     * Retrieves the link to submit request to delete the session.
+     * @param courseId course ID
+     * @param feedbackSessionName the session name
+     * @param isHome true to indicate the request is from the home page
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/PageData.java]: This javadoc parameter should be updated I guess

File: src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java (hunk 3 above)
Comment (kennho) [inline src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java]: oh. this is to set the isHome to true so the correct return url can be generated at the InstructorFeedbackSessionActions.java

**Comment by kanghj** [comment]
Let's wait for @yj-soh to give comments on #5089, since the three PRs you have are roughly similar

**Comment by kennho** [comment]
Okay. Updated the comments too @kanghj

**Comment by thyageshm** [comment]
@kanghj ready for review

File: src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java [hunk 5]
```diff
@@ -237,23 +237,20 @@ private void testCreateInstructorAction() throws Exception {
         
         ______TS("new instructor can send reminder of sample course");
         instructorHomePage.loadInstructorHomeTab();
-        feedbacksPage = instructorHomePage.clickFeedbackSessionRemindLink("AHPUiT.instr1.gma-demo", 
-                                                                          "Second team feedback session");
+        feedbacksPage = instructorHomePage.clickFeedbackSessionRemindLink("AHPUiT.instr1.gma-demo", "Second team feedback session");
```
Comment (yj-soh) [inline src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java]: Is there a reason for removing the line break for the sake of shortening the lines?

File: src/main/java/teammates/ui/template/FeedbackSessionPublishButton.java [hunk 6]
```diff
@@ -25,6 +25,8 @@ public FeedbackSessionPublishButton(PageData data, FeedbackSessionAttributes ses
         this.feedbackSessionName = Sanitizer.sanitizeForJs(session.feedbackSessionName);
         this.isSendingPublishedEmail = session.isPublishedEmailEnabled;
 
+        String returnUrl = isHome ? Const.ActionURIs.INSTRUCTOR_HOME_PAGE 
+                                  : Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE;
```
Comment (yj-soh) [inline src/main/java/teammates/ui/template/FeedbackSessionPublishButton.java]: Is there a reason for passing in `isHome` as a parameter rather than `returnUrl`?

Comment (kennho) [inline src/main/java/teammates/ui/template/FeedbackSessionPublishButton.java]: opps sorry left out this part! will make the necessary changes

**Comment by kennho** [comment]
Updated with suggested changes @yj-soh @kanghj

File: src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java [hunk 7]
```diff
@@ -34,6 +34,8 @@ public InstructorFeedbackSessionActions(PageData data, FeedbackSessionAttributes
                                             InstructorAttributes instructor) {
         String courseId = session.courseId;
         String feedbackSessionName = session.feedbackSessionName;
+        String returnUrl = isHome ? Const.ActionURIs.INSTRUCTOR_HOME_PAGE
+                                  : Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE;
```
Comment (yj-soh) [inline src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java]: Actually, if we want extensibility, it's probably best to just use `returnUrl` from the start, with the `*PageData` calling such functions and passing in the URI rather than a boolean like `isHome`.

File: src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java [hunk 8]
```diff
@@ -61,8 +61,7 @@ public InstructorFeedbackSessionActions(PageData data, FeedbackSessionAttributes
         this.allowedToSubmit = (session.isVisible() || session.isPrivateSession()) && shouldEnableSubmitLink;
         this.allowedToRemind = session.isOpened() && instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION);
             
-        this.publishButton = new FeedbackSessionPublishButton(data, session, isHome, instructor,
-                                                              PUBLISH_BUTTON_TYPE);
+        this.publishButton = new FeedbackSessionPublishButton(data, session, returnUrl, instructor,PUBLISH_BUTTON_TYPE);
```
Comment (yj-soh) [inline src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java]: Fix the formatting here.
Keep to our line length limits wherever possible.

File: src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java [hunk 9]
```diff
@@ -237,23 +237,20 @@ private void testCreateInstructorAction() throws Exception {
...
         ______TS("new instructor can publish feedbackSession of sample course");
         instructorHomePage.loadInstructorHomeTab();
-        feedbacksPage = instructorHomePage.clickFeedbackSessionPublishLink("AHPUiT.instr1.gma-demo", 
-                                                                           "Second team feedback session");
-        feedbacksPage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionPublished.html");
+        instructorHomePage.clickFeedbackSessionPublishLink("AHPUiT.instr1.gma-demo", "Second team feedback session");
+        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionPublished.html");
         
-        feedbacksPage.logout();
+        instructorHomePage.logout();
```
Comment (yj-soh) [inline src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java]: Does `feedbacksPage` need to be logged out?

Comment (kennho) [inline src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java]: Hmm. I think they both calls the same method in AppData.java which is to click the logout button. Tried using either one. Both does the same thing of logging out.

**Comment by kennho** [comment]
Updated with suggested changes @kanghj @yj-soh

**Comment by yj-soh** [comment]
It looks okay to me, @kanghj, anything from you?

I do have some general comments on the PR, though.
When updating tests, why not put them in a single commit if they are related?
Having different commits for each file only makes it harder to trace the exact cause in the future and messes up the repo history.
The comment of "Updated test pages affected by {name}" is also pretty worthless when you don't actually mention why.

When merging in master to update the branch, try keeping that on your local repo and clean it up before pushing.
There's absolutely no need to have 10 consequent commits for merging in master.
Or rather, there's no need to even merge in master unless you require something new that is available only in master.
Those commits will end up in master if the branch gets merged, so all those unnecessary merge commits will end up polluting the master branch as well.

**Comment by kennho** [comment]
Oops! Sorry! It is stated to sync with the committer repo frequently so I did that everytime there is an update to the master 😅  Will take note of it in the future. @yj-soh

**Comment by damithc** [comment]
Get rid of noisy commits?

**Comment by kennho** [comment]
do I still remove the multiple commits? I am currently trying to rebase and squash all the merge master commits into one and the update tests commits into one @kanghj

**Comment by kanghj** [comment]
Yup, just do it

**Comment by kennho** [comment]
Done removing @kanghj @yj-soh
